### PR TITLE
sdk: give the ability to get a room's preview

### LINKED ISSUE #3298: [Permalink] Retrieve information about a Room.
When a permalink is detected in a message, the application needs an API on the Client interface to get information about what's behind the link. It can be useful for rendering, like for instance converting:

https://matrix.to/#/!aCvOWKNDeXXpPgquKq:matrix.org/$yWTULhbxy57KcXc5vjewFz74Myr5-LmLOx4WebPDkeQ?via=matrix.org&via=envs.net&via=tchncs.de 

to 

image

But also (and as a priority) useful to handle the Room preview (if not joined or invited) and the navigation in the app

From a matrix.to link to a room or an event (or whatever the base URL is, see [this comment](https://github.com/matrix-org/matrix-rust-sdk/issues/3259#issuecomment-2036561128)) and so from the roomId or the room alias, we will need this information:

- roomId: String
- alias: String?
- name: String?
- topic: String?
- avatarUrl: String?
- numJoinedMembers: Int?
- roomType: String?
- viaServers: List<String>
- isPublic: Boolean
- <del>isSpace: Boolean</del> *Not necessary, redundant with `roomType`*
- isJoined: Boolean
- isUserInvited: Boolean
- canKnock: Boolean

If not possible the parsing and eventually alias resolving should provide those informations
 - roomId: String,
 - alias: String?,
 - viaServers: List<String>

If the room alias is not known and cannot be resolved an error should be returned.

The code in the Kotlin SDK can be found [here](https://github.com/element-hq/element-android/blob/b3fdb5ed17d057003581bebefd64352e9c3abbd2/matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/peeking/PeekRoomTask.kt) 

Also see the design here: https://www.figma.com/file/Ni6Ii8YKtmXCKYNE90cC67/Timeline-(new)?type=design&node-id=559-3189&mode=design&t=zBw1sSd7rXLCrxSY-0

### PR DESCRIPTION
**Pull request by bnjbvr**
Fixes #3298.

### CODE REVIEW

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 1]
```diff
@@ -1145,14 +1154,20 @@ impl RoomInfo {
         }
     }
 
-    fn history_visibility(&self) -> &HistoryVisibility {
+    /// Returns the history visiblity for this room.
```
Comment (andybalaam) [inline crates/matrix-sdk-base/src/rooms/normal.rs:1157]:
```suggestion
    /// Returns the history visibility for this room.
```

File: crates/matrix-sdk/src/room_preview.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    /// The number of joined members.
+    pub num_joined_members: u64,
+
+    /// The room type (space, custom) or nothing, if it's a regular room.
+    pub room_type: Option<RoomType>,
+
+    /// What's the join rule for this room?
+    pub join_rule: JoinRule,
+
+    /// What's the history visiblity for this room?
```
Comment (andybalaam) [inline crates/matrix-sdk/src/room_preview.rs:57]:
```suggestion
    /// What's the history visibility for this room?
```

**Review by andybalaam — APPROVED** [review]
Looks good, thanks! (2 typos)

**Comment by bnjbvr** [comment]
Thanks for the quick review! I'll merge this, as there are already a few improvements that have been requested (a better error type for rooms one can't preview, and a way to preview public rooms via the room directory)

**Comment by stefanceriu** [comment]
@bnjbvr can you have a look at this please? seems one of the tests failed

```
2024-04-16T16:02:24.955962Z DEBUG sync_once{pos="2"}:send{server_versions=[V1_0, V1_1, V1_2, V1_3, V1_4, V1_5, V1_6] config=RequestConfig { timeout: 60s } request_id="REQ-18" method=POST uri="http://localhost:8118/_matrix/client/unstable/org.matrix.msc3575/sync" request_size="129 B" request_body=b"{\"conn_id\":\"room-list\",\"lists\":{\"all_rooms\":{\"ranges\":[[0,0]]},\"invites\":{\"ranges\":[[0,0]]},\"visible_rooms\":{\"ranges\":[[0,19]]}}}"}: matrix_sdk::http_client: Sending request
thread 'tests::sliding_sync::room::test_room_preview' panicked at testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs:1129:40:
called `Option::unwrap()` on a `None` value


failures:
    tests::sliding_sync::room::test_room_preview
```

**Pull request merged**